trino-aws-proxy-spark3/src/main/java/io/trino/aws/proxy/spark3/PresignAwareAmazonS3.java

```diff
@@ -0,0 +1,1667 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+class PresignAwareAmazonS3
+        implements AmazonS3
+{
+    private static final Log log = LogFactory.getLog(PresignAwareAmazonS3.class);
+
+    private final AmazonS3 delegate;
+
+    PresignAwareAmazonS3(AmazonS3 delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    private record Presigned(URL url, ObjectMetadata objectMetadata)
+    {
+        private Presigned
+        {
+            requireNonNull(url, "url is null");
+            requireNonNull(objectMetadata, "objectMetadata is null");
+        }
+    }
+
+    private Optional<Presigned> getPresignedUrl(String method, String bucket, String key)
+    {
+        return getPresignedUrl(method, bucket, key, Optional.empty());
+    }
+
+    private Optional<Presigned> getPresignedUrl(String method, String bucket, String key, Optional<UploadPartRequest> uploadPartRequest)
+    {
+        Optional<String> uploadId = uploadPartRequest.map(UploadPartRequest::getUploadId);
+        Optional<Integer> partNumber = uploadPartRequest.map(UploadPartRequest::getPartNumber);
+
+        log.debug("request.get-presigned-url method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber));
+
+        Map<String, ?> headers;
+        ObjectMetadata objectMetadata;
+        try {
+            GetObjectMetadataRequest getObjectMetadataRequest = new GetObjectMetadataRequest(bucket, key);
+            partNumber.ifPresent(getObjectMetadataRequest::setPartNumber);
+            uploadId.ifPresent(id -> getObjectMetadataRequest.putCustomQueryParameter("uploadId", id));
+
+            objectMetadata = delegate.getObjectMetadata(getObjectMetadataRequest);
+            headers = objectMetadata.getRawMetadata();
+        }
+        catch (AmazonServiceException e) {
+            headers = e.getHttpHeaders();
+            ObjectMetadata work = new ObjectMetadata();
+            headers.forEach(work::setHeader);
+            objectMetadata = work;
+        }
+        catch (RuntimeException e) {
+            log.error("request.get-presigned-url.exception method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber), e);
+            throw e;
+        }
+
+        Object value = headers.get("X-Trino-Pre-Signed-Url-" + method);
+        log.debug("response.get-presigned-url.exception method=%s bucket=%s key=%s uploadId=%s partNumber=%s success=%s".formatted(method, bucket, key, uploadId, partNumber, value != null));
+        try {
+            return (value != null) ? Optional.of(new Presigned(new URL(String.valueOf(value)), objectMetadata)) : Optional.empty();
```

Comment:
Sadly, Java is letting us down here. a) `new URL` throws an exception so we'd have to wrap that. b) `objectMetadata` is not effectively final so we'd have to wrap that. This makes the elegant line of code you have above not-so elegant. We could add some helper methods to fix it but it's probably better to leave it as is.

Reply:
Make sense, I was expecting there was a reason you didn't use `map` in the first place. 

Could we make it an `if` statement instead of a ternary op?

Reply:
What's wrong with ternary? I like the one-liner.

Reply:
I guess its just preference, I like the verbosity of if for returns in particular. I'm fine with keeping it like this